apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -72,6 +81,44 @@ store_message(Msg) ->
 find_subscribers(_Msg) ->
     [node()].
 
+open_session(ClientID) ->
+    ?WHEN_ENABLED(emqx_ds:session_open(ClientID)).
+
+-spec add_subscription(emqx_types:topic(), emqx_ds:session_id()) ->
+    {ok, emqx_ds:iterator_id(), _IsNew :: boolean()} | {skipped, disabled}.
+add_subscription(TopicFilterBin, DSSessionID) ->
+    ?WHEN_ENABLED(
+        begin
+            TopicFilter = emqx_topic:words(TopicFilterBin),
+            {ok, IteratorID, StartMS, IsNew} = emqx_ds:session_add_iterator(
+                DSSessionID, TopicFilter
+            ),
+            case IsNew of
+                true ->
+                    ok = open_iterator_on_all_nodes(TopicFilter, StartMS, IteratorID);
+                false ->
+                    ok
+            end,
```

Comment:
Note that this whole operation (`add_subscription/2`) is not atomic. We should be careful to make sure it behaves "idempotently" even if, for example, the node died after the successful `eqmx_ds:session_add_iterator/2` operation.

In general, with the code that touches some stable storage it's probably worth thinking: what if the node unexpectedly dies here? To minimize the chance we'll end up with subtly broken state.